test/common/upstream/health_check_fuzz.cc

```diff
@@ -112,34 +111,402 @@ void HealthCheckFuzz::triggerTimeoutTimer(bool last_action) {
+      break;
+    }
+    case test::common::upstream::ServingStatus::SERVING: {
+      data = serializeResponse(grpc::health::v1::HealthCheckResponse::SERVING);
+      break;
+    }
+    case test::common::upstream::ServingStatus::NOT_SERVING: {
+      data = serializeResponse(grpc::health::v1::HealthCheckResponse::NOT_SERVING);
+      break;
+    }
+    case test::common::upstream::ServingStatus::SERVICE_UNKNOWN: {
+      data = serializeResponse(grpc::health::v1::HealthCheckResponse::SERVICE_UNKNOWN);
+      break;
+    }
+    default: { // shouldn't hit
+      NOT_REACHED_GCOVR_EXCL_LINE;
+      break;
+    }
+    }
+    Buffer::OwnedImpl response;
+    response.add(data.data(), data.size());
+    return response;
+  }
+  case test::common::upstream::GrpcRespondBytes::kData: {
+    // Arbitrarily Generated Bytes
+    Buffer::OwnedImpl response;
+    response.add(grpc_respond_bytes.data());
+
+    ENVOY_LOG_MISC(trace,
+                   "Responded with {}. Length (in bytes) = {}. These bytes were arbitrarily "
+                   "generated and have no structure.",
+                   grpc_respond_bytes.data(), grpc_respond_bytes.data().length());
+
+    return response;
+  }
+  default: { // shouldn't hit
+    NOT_REACHED_GCOVR_EXCL_LINE;
+    break;
+  }
+  }
+}
+
+// Logic from respondResponseSpec() in unit tests
+void GrpcHealthCheckFuzz::respond(test::common::upstream::GrpcRespond grpc_respond) {
+  if (!test_sessions_[0]->timeout_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Timeout timer is disabled. Skipping response.");
+    return;
+  }
+  // These booleans help figure out when to end the stream
+  bool has_data = false;
+  bool has_trailers = false;
+
+  if (grpc_respond.has_grpc_respond_bytes()) {
+    switch (grpc_respond.grpc_respond_bytes().grpc_respond_bytes_selector_case()) {
+    case test::common::upstream::GrpcRespondBytes::kStatus: {
+      has_data = true;
+      break;
+    }
+    case test::common::upstream::GrpcRespondBytes::kData: {
+      if (grpc_respond.grpc_respond_bytes().data().length() != 0) {
+        has_data = true;
+      }
+      break;
+    }
+    default: // If the fuzzer generates an empty grpc_respond_bytes message
+      break;
+    }
+  }
+  // TODO: Should I hardcode trailers? Particularly grpc-status = 0?
+  if (grpc_respond.has_grpc_respond_trailers()) {
+    if (grpc_respond.grpc_respond_trailers().has_trailers()) {
+      if (grpc_respond.grpc_respond_trailers().trailers().headers_size() != 0) {
+        has_trailers = true;
+      }
+    }
+  }
+
+  ENVOY_LOG_MISC(trace, "Has data: {}. Has trailers: {}.", has_data, has_trailers);
+
+  bool end_stream_on_headers = !has_data && !has_trailers;
+
+  std::unique_ptr<Http::TestResponseHeaderMapImpl> response_headers =
+      std::make_unique<Http::TestResponseHeaderMapImpl>(
+          Fuzz::fromHeaders<Http::TestResponseHeaderMapImpl>(
+              grpc_respond.grpc_respond_headers().headers(), {}, {}));
```

Comment:
what if there are no headers in the GrpcRespond (since it's not required)? does that mean you'll call decodeHeaders with an empty headermap, and then continue the logic below?

Reply:
No, I expiclity add a hardcoded status header as 200.

Reply:
Line right below this one :)